### PR TITLE
Add promise support to built-in model ACL

### PR DESCRIPTION
**Pull request by ebarault**
### Description

None of the functions from built-in acl.js currently support promises. this PR adds promise support by using loopback utils helper
 
#### Related issues
related issue: https://github.com/strongloop/loopback/issues/418
related PR (similar for built-in role model) : https://github.com/strongloop/loopback/pull/3146

- None

### Checklist

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by ebarault** [comment]
@bajtos @superkhau please review so we can land this before https://github.com/strongloop/loopback/pull/2971

File: common/models/acl.js [hunk 1]
```diff
@@ -546,16 +544,21 @@ module.exports = function(ACL) {
...
    */
   ACL.isMappedToRole = function(principalType, principalId, role, cb) {
+    cb = cb || utils.createPromiseCallback();
     var self = this;
     this.resolvePrincipal(principalType, principalId,
       function(err, principal) {
-        if (err) return cb(err);
+        if (err) {
+          return cb(err);
+        }
```
Comment (superkhau) [inline common/models/acl.js]: this change is not required

File: common/models/acl.js [hunk 2]
```diff
@@ -546,16 +544,21 @@ module.exports = function(ACL) {
...
         if (principal != null) {
           principalId = principal.id;
         }
         principalType = principalType || 'ROLE';
         self.resolvePrincipal('ROLE', role, function(err, role) {
-          if (err || !role) return cb(err, role);
+          if (err || !role) {
+            return cb(err, role);
+          }
```
Comment (superkhau) [inline common/models/acl.js]: ditto

File: test/acl.test.js [hunk 3]
```diff
@@ -90,6 +90,42 @@ describe('security scopes', function() {
 });
 
 describe('security ACLs', function() {
+  it('supports checkPermission() returning a promise', function(done) {
+    ACL.create({
```
Comment (superkhau) [inline test/acl.test.js]:
```
return ACL.create({
    principalType: ACL.USER, principalId: 'u001', model: 'testModel', property: ACL.ALL,
    accessType: ACL.ALL, permission: ACL.ALLOW
  })
  .then(function() {
    return ACL.checkPermission(ACL.USER, 'u001', 'testModel', 'name', ACL.ALL);
  })
.then(function(access) {
  assert(access.permission === ACL.ALLOW);
});

File: test/acl.test.js [hunk 4]
```diff
@@ -90,6 +90,42 @@ describe('security scopes', function() {
...
+      return ACL.checkPermission(ACL.USER, 'u001', 'testModel', 'name', ACL.ALL);
+    })
+    .then(function(access) {
+      assert(access.permission === ACL.ALLOW);
+      done();
+    })
+    .catch(done);
+  });
+
+  it('should support checkAccessForContext() returning a promise', function(done) {
```
Comment (superkhau) [inline test/acl.test.js]: remove should

File: test/acl.test.js [hunk 5]
```diff
@@ -90,6 +90,42 @@ describe('security scopes', function() {
...
+    var testModel = ds.createModel('testModel', {
+      acls: [
+        {principalType: ACL.USER, principalId: 'u001',
+          accessType: ACL.ALL, permission: ACL.ALLOW},
+      ],
+    });
+
+    ACL.checkAccessForContext({
+      principals: [
+          {type: ACL.USER, id: 'u001'},
```
Comment (superkhau) [inline test/acl.test.js]: indent is off

File: test/acl.test.js [hunk 6]
```diff
@@ -90,6 +90,42 @@ describe('security scopes', function() {
...
+
+  it('should support checkAccessForContext() returning a promise', function(done) {
+    var testModel = ds.createModel('testModel', {
+      acls: [
+        {principalType: ACL.USER, principalId: 'u001',
+          accessType: ACL.ALL, permission: ACL.ALLOW},
+      ],
+    });
+
+    ACL.checkAccessForContext({
```
Comment (superkhau) [inline test/acl.test.js]: return ACL.checkAccess...

File: test/acl.test.js [hunk 7]
```diff
@@ -90,6 +90,42 @@ describe('security scopes', function() {
...
+      principals: [
+          {type: ACL.USER, id: 'u001'},
+      ],
+      model: 'testModel',
+      accessType: ACL.ALL,
+    })
+    .then(function(access) {
+      assert(access.permission === ACL.ALLOW);
```
Comment (superkhau) [inline test/acl.test.js]: remove done

File: test/role.test.js [hunk 8]
```diff
@@ -531,6 +531,15 @@ describe('role model', function() {
       });
     });
 
+    it('should support ACL.resolvePrincipal() returning a promise', function(done) {
+      ACL.resolvePrincipal(ACL.USER, user.id)
+      .then(function(u) {
+        expect(u.id).to.eql(user.id);
+        done();
+      })
+      .catch(done);
```
Comment (superkhau) [inline test/role.test.js]: remove catch

File: test/acl.test.js (hunk 3 above)
Comment (superkhau) [inline test/acl.test.js]: See http://loopback.io/doc/en/contrib/style-guide.html for more info on promise style tests in mocha

File: test/role.test.js [hunk 9]
```diff
@@ -531,6 +531,15 @@ describe('role model', function() {
       });
     });
 
+    it('should support ACL.resolvePrincipal() returning a promise', function(done) {
```
Comment (superkhau) [inline test/role.test.js]: remove should and done

File: test/acl.test.js [hunk 10]
```diff
@@ -90,6 +90,42 @@ describe('security scopes', function() {
...
+      done();
+    })
+    .catch(done);
```
Comment (superkhau) [inline test/acl.test.js]: remove catch

File: test/role.test.js [hunk 11]
```diff
@@ -531,6 +531,15 @@ describe('role model', function() {
...
+      ACL.resolvePrincipal(ACL.USER, user.id)
+      .then(function(u) {
+        expect(u.id).to.eql(user.id);
+        done();
```
Comment (superkhau) [inline test/role.test.js]: remove done

File: test/role.test.js [hunk 12]
```diff
@@ -581,6 +590,15 @@ describe('role model', function() {
       });
     });
 
+    it('should support ACL.isMappedToRole() returning a promise', function(done) {
```
Comment (superkhau) [inline test/role.test.js]: remove should and done

File: test/role.test.js [hunk 13]
```diff
@@ -581,6 +590,15 @@ describe('role model', function() {
...
+      ACL.isMappedToRole(ACL.USER, user.username, 'admin')
```
Comment (superkhau) [inline test/role.test.js]: return ACL.isMapped...

File: test/role.test.js [hunk 14]
```diff
@@ -581,6 +590,15 @@ describe('role model', function() {
...
+      .then(function(flag) {
+        expect(flag).to.eql(true);
+        done();
```
Comment (superkhau) [inline test/role.test.js]: remove done

File: test/role.test.js [hunk 15]
```diff
@@ -581,6 +590,15 @@ describe('role model', function() {
       });
     });
 
+    it('should support ACL.isMappedToRole() returning a promise', function(done) {
+      ACL.isMappedToRole(ACL.USER, user.username, 'admin')
+      .then(function(flag) {
+        expect(flag).to.eql(true);
```
Comment (superkhau) [inline test/role.test.js]: `expect(flag).to.be.true();`

File: test/role.test.js [hunk 16]
```diff
@@ -581,6 +590,15 @@ describe('role model', function() {
...
+        done();
+      })
+      .catch(done);
```
Comment (superkhau) [inline test/role.test.js]: remove catch

File: test/acl.test.js (hunk 3 above)
Comment (ebarault) [inline test/acl.test.js]: hi @superkhau : i am aware of this, i thought you'd like to stick to the origin test file style. as per recent @bajtos PR where the callback style is still used (e.g. [here in `role.test.js`](https://github.com/strongloop/loopback/pull/3146/files#diff-b7e36d904fe1d290934c235affb649ecR308))
i'll modify to use mocha with promises

**Comment by ebarault** [comment]
@superkhau : ready for review

**Comment by superkhau** [comment]
@ebarault Thanks for the contribution. LGTM. @bajtos For final review/input though.

File: test/acl.test.js (hunk 3 above)
Comment (bajtos) [inline test/acl.test.js]: I vaguely remember there was a reason for using callback-style tests when verifying Promise-based API, but I cannot recall what was it. As I was reviewing the code in this patch, I didn't see why the new style should not work, so let's stick to what our style guide says.

**Comment by bajtos** [comment]
@superkhau thank you for the detailed review!

**Comment by bajtos** [comment]
@slnode test please

**Comment by bajtos** [comment]
@slnode test please

**Comment by bajtos** [comment]
Landed, thank you for the contribution!

And sorry for the comment spam, our CI was acting up a bit today 😕

**Pull request merged**